tests/core/container/test_core_avl.odin

```diff
@@ -0,0 +1,158 @@
+package test_core_container
+
+import "core:container/avl"
+import "core:math/rand"
+import "core:slice"
+import "core:testing"
+
+import tc "tests:common"
+
+@(test)
+test_avl :: proc(t: ^testing.T) {
+	tc.log(t, "Testing avl")
+
+	// Initialization.
+	tree: avl.Tree(int)
+	avl.init(&tree, proc(a, b: int) -> int {
+		return a - b
+	})
+	tc.expect(t, avl.len(&tree) == 0, "empty: len should be 0")
+	tc.expect(t, avl.first(&tree) == nil, "empty: first should be nil")
+	tc.expect(t, avl.last(&tree) == nil, "empty: last should be nil")
+
+	iter := avl.iterator(&tree, avl.Direction.Forward)
+	tc.expect(t, avl.iterator_first(&iter) == nil, "empty/iterator: first should be nil")
+
+	// Test insertion.
+	NR_INSERTS :: 32 + 1 // Ensure at least 1 collision.
+	insertedMap := make(map[int]^avl.Node(int))
+	for i := 0; i < NR_INSERTS; i += 1 {
+		v := int(rand.uint32() & 0x1f)
+		existingNode, inMap := insertedMap[v]
+
+		ok, n, _ := avl.insert(&tree, v)
+		tc.expect(t, inMap != ok, "insert: ok should match inverse of map lookup")
+		if ok {
+			insertedMap[v] = n
+		} else {
+			tc.expect(t, existingNode == n, "insert: expecting existing node")
+		}
+	}
+	nrEntries := len(insertedMap)
+	tc.expect(t, avl.len(&tree) == nrEntries, "insert: len after")
+	tree_validate(t, &tree)
+
+	// Ensure that all entries can be found.
+	for k, v in insertedMap {
+		tc.expect(t, v == avl.find(&tree, k), "Find(): Node")
+		tc.expect(t, k == v.value, "Find(): Node value")
+	}
+
+	// Test the forward/backward iterators.
+	insertedValues: [dynamic]int
+	for k in insertedMap {
+		append(&insertedValues, k)
+	}
+	slice.sort(insertedValues[:])
+
+	iter = avl.iterator(&tree, avl.Direction.Forward)
+	visited: int
+	for node := avl.iterator_first(&iter); node != nil; node = avl.iterator_next(&iter) {
```

Comment:
When you make `iterator_next` return a `bool` and make `iterator` initialize itself, you can just make this `for node := iterator_next(&iter) {`